### PR TITLE
Fix a crash in `_WritePseudoWindowCallback`

### PR DESCRIPTION
**Pull request by zadjii-msft**
Fixes MSFT:40853556

There's a small race here. The renderer thread in ConPTY might notice the terminal is gone, call CloseOutput, and release the vt renderer, and then the window proc fires and decides to minimize/restore the window, triggering an A/V.

I'm 100% confident that this has NEVER happened to a real user. But the test labs hit it so much that it makes up ~26% of our crashes.

I haven't tested this cause again, _it doesn't hit in the wild_

### CODE REVIEW

**Comment by DHowett** [comment]
Could this also be MSFT:40913882?

**Review by DHowett — APPROVED** [review]

**Comment by DHowett** [comment]
LOL, thanks for fixing my fixe. I did the same thing to a contributor's PR this exact same week.

**Comment by zadjii-msft** [comment]
> Could this also be [MSFT:40913882](https://task.ms/40913882)?

Nah, I'll reply with investigation in that bug.

File: src/host/VtIo.cpp [hunk 1]
```diff
@@ -258,8 +258,20 @@ bool VtIo::IsUsingVt() const
...
-                LOG_IF_FAILED(_pVtRenderEngine->SetWindowVisibility(showOrHide));
+                // MSFT:40853556 Grab the shutdown lock here, so that another
+                // thread can't trigger a CloseOutput and release the
+                // _pVtRenderEngine out from underneath us.
+                //
+                // I'm doing this instead of just
+                // SetPseudoWindowCallback(nullptr), so that we don't just move
+                // the A/V race to in between checking
+                // _pseudoWindowMessageCallback and actually calling it.
+                std::lock_guard<std::mutex> lk(_shutdownLock);
```
Comment (lhecker) [inline src/host/VtIo.cpp:269]: FYI Since C++17 you don't need to specify template parameters anymore - they'll be automatically inferred.

**Review by lhecker — APPROVED** [review]

**Comment by lhecker** [comment]
Just from reading the diff, I don't quite get why we wouldn't have to add these guards everywhere else. What makes `SetWindowVisibility` so special here?

**Comment by zadjii-msft** [comment]
> Just from reading the diff, I don't quite get why we wouldn't have to add these guards everywhere else. What makes `SetWindowVisibility` so special here?

Presumably, this is the only bit of code that can be called from another thread in an unsafe way here. Typically everything else we were doing was always during startup or immediately on shutdown, but now we've got this pesky user32 input thread to deal with window messages that can get in here too now

**Comment by DHowett** [comment]
That explanation makes sense to me!

**Comment by DHowett** [comment]
@msftbot merge this in 5 minutes

**Pull request merged**